enhancements/20221116-sriov-vf-provisioner.md

```diff
@@ -0,0 +1,91 @@
+    note right of H: (Different SR-IOV drivers have their own way to create the VF)
+    H ->> G: vf-provisioner hotplugs the new VF as a raw PCI Device
+```
+
+The last step, the hot-plugging into the VM's container is to allow for dynamically adding the device to the virt-launcher container.
+
+### Related Issues
+
+- https://github.com/harvester/harvester/issues/2763
+
+## Motivation
+
+In order to keep track of request for SR-IOV virtual functions (vfs), the vf-provisioner is proposed.
+Currently, the pcidevices controller can assign vfs to VMs, but it cannot install the drivers that allow the vfs to be created.
+Also the details of how to create the vf are driver-specific, so we need something that has that knowledge baked in.
+This design is loosely inspired by the csi-provisioner architecture.
+
+### Goals
+
+- Allow guest cluster to request a VF to be allocated to a pod or a vm
+- Modify the host Harvester cluster to install the SR-IOV drivers that expose the ability to create VFs
+- Track which created VFs have been allocated
+- Allow hot-plugging VFs into running Pods
+
+## Proposal
+
+
+
+### User Stories
+Detail the things that people will be able to do if this enhancement is implemented. A good practise is including a comparsion of what user cannot do before the enhancement implemented, why user would want an enhancement and what user need to do after, to make it clear why the enhancement beneficial to the user.
+
+The experience details should be in the `User Experience In Detail` later.
+
+#### Story 1
+Joe works for a Telco and has a long-running VM with several VFs currently reserved, serving various containers in the guest cluster. There is spare capacity on the VM and a call comes in. He wants the VM to be able to add the VF without having to restart the VM. Using [CNI](https://www.cni.dev/), the virt-launcher's PodSpec would have to be changed, which would require the VM to be rebooted. In order to avoid this, the VF will be hotplugged into the running VM using `qemu`'s [PCI hotplugging capability](https://www.linux-kvm.org/page/Hotadd_pci_devices#Using_PCI_Hotplug_Support).
+
+
+### API changes
+
+- Adding the VFClaim CRD to the host and guest clusters
+- Adding an vf-provisioner with an endpoint to request
+    - VFs be created and claimed for a VM
+    - VFs to be unclaimed and returned to a pool on the host
+    - VFs to be destroyed (decrement the num_vfs value on the host's SR-IOV driver)
+
+## Design
+
+### Implementation Overview
+
+Create a VFClaim that stores:
+- Guest node on which the VM is running
+- The VM IP 
+- The host node on which the VF's PF lives
+- The VF's PCI address on the node (when the guest creates this, it will initially be blank, then it is replicated to the host, where it's populated, and then the guest cluster syncs that value eventually. Discuss: Is this necessary?)
+
+Create a controller called vf-provisioner, that will run on the host Harvester cluster.
```

Comment:
```suggestion
Create a VFClaim that stores:
- The guest node on which the VM is running.
- The VM IP.
- The host node on which the VF's PF lives.
- The VF's PCI address on the node. When the guest creates this, it will initially be blank. Then it is replicated to the host, where it's populated, and the guest cluster eventually syncs that value. (Discuss: Is this necessary?)

Create a controller called vf-provisioner that will run on the host Harvester cluster.
```